src/main/java/com/spotify/docker/client/auth/ecr/ContainerRegistryAuthSupplier.java

```diff
@@ -0,0 +1,233 @@
+import com.spotify.docker.client.auth.RegistryAuthSupplier;
+import com.spotify.docker.client.exceptions.DockerException;
+import com.spotify.docker.client.messages.RegistryAuth;
+import com.spotify.docker.client.messages.RegistryConfigs;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A RegistryAuthSupplier for authenticating against an AWS Elastic Container Registry.
+ */
+public class ContainerRegistryAuthSupplier implements RegistryAuthSupplier {
+
+  private static final Logger log = LoggerFactory.getLogger(ContainerRegistryAuthSupplier.class);
+
+  /**
+   * Constructs a ContainerRegistryAuthSupplier using the Application Default Credentials.
+   *
+   * @see Builder
+   */
+  public static Builder forDefaultClient() throws IOException {
+    return new Builder(AmazonECRClientBuilder.defaultClient());
+  }
+
+  /**
+   * Constructs a ContainerRegistryAuthSupplier using the specified credentials.
+   *
+   * @see Builder
+   */
+  public static Builder forEcrClient(final AmazonECR ecr) {
+    return new Builder(ecr);
+  }
+
+  /**
+   * A Builder of ContainerRegistryAuthSupplier.
+   * <p>
+   * The default value for the minimum expiry time of an access token is one minute. When the
+   * ContainerRegistryAuthSupplier is asked for a RegistryAuth, it will check if the existing
+   * authorization token for the AWS AuthorizationData expires within this amount of time.
+   * If it does, then the AuthorizationData is refreshed before being returned.
+   * </p>
+   */
+  public static class Builder {
+
+    private final AmazonECR ecr;
+    private long minimumExpiryMillis = TimeUnit.MINUTES.toMillis(1);
+
+    public Builder(final AmazonECR ecr) {
+      this.ecr = ecr;
+    }
+
+    /**
+     * Changes the minimum expiry time used to refresh AccessTokens before they expire. The default
+     * value is one minute.
+     */
+    public Builder withMinimumExpiry(long duration, TimeUnit timeUnit) {
+      this.minimumExpiryMillis = TimeUnit.MILLISECONDS.convert(duration, timeUnit);
+
+      return this;
+    }
+
+    public ContainerRegistryAuthSupplier build() {
+      final Clock clock = Clock.SYSTEM;
+
+      return new ContainerRegistryAuthSupplier(ecr, clock, minimumExpiryMillis,
+          new EcrCredentials(ecr));
+    }
+  }
+
+  private final AmazonECR ecr;
+
+  // TODO (mbrown): change to java.time.Clock once on Java 8
+  private final Clock clock;
+  private EcrCredentials credentials;
+  private final long minimumExpiryMillis;
+
+  @VisibleForTesting
+  ContainerRegistryAuthSupplier(final AmazonECR ecr, final Clock clock,
+                                final long minimumExpiryMillis, final EcrCredentials credentials) {
+    Preconditions.checkArgument(ecr != null, "ecr");
+    this.ecr = ecr;
```

Comment:
Do we need to pass `ecr` here? I see it's used as a sync object below, but can't you just lock on something else?